source/extensions/filters/http/header_to_metadata/header_to_metadata_filter.cc

```diff
@@ -164,15 +172,32 @@ void HeaderToMetadataFilter::writeHeaderToMetadata(Http::HeaderMap& headers,
                                                    Http::StreamFilterCallbacks& callbacks) {
   StructMap structs_by_namespace;
 
-  for (const auto& rulePair : rules) {
-    const auto& header = rulePair.first;
-    const auto& rule = rulePair.second;
+  for (const auto& rule : rules) {
+    const auto& header = rule.header();
+    const auto& proto_rule = rule.rule();
     const Http::HeaderEntry* header_entry = headers.get(header);
 
-    if (header_entry != nullptr && rule.has_on_header_present()) {
-      const auto& keyval = rule.on_header_present();
-      absl::string_view value = keyval.value().empty() ? header_entry->value().getStringView()
-                                                       : absl::string_view(keyval.value());
+    if (header_entry != nullptr && proto_rule.has_on_header_present()) {
+      const auto& keyval = proto_rule.on_header_present();
+      absl::string_view value = header_entry->value().getStringView();
+      // This is used to hold the rewritten header value, so that it can
+      // be bound to value without going out of scope.
+      std::string rewritten_value;
+
+      if (!keyval.value().empty()) {
+        value = absl::string_view(keyval.value());
+      } else {
+        const auto& matcher = rule.regexRewrite();
+        if (matcher != nullptr) {
+          rewritten_value = matcher->replaceAll(value, rule.regexSubstitution());
+          // If there was no replacement, don't add the metadata.
```

Comment:
Hmm. I can't think of case in which a match + substitution = no-op would be useful. Any examples? Also note that allowing no-op substitutions (as is) also means no matches are allowed too. Which in my use case, would break my subset LB config (I use `NO_FALLBACK` policy).

If we do want to support no-op substitutions and/or no match also triggers adding metadata, we'd probably want it to be the non default behavior (`bool no_rewrite_adds_metadata`?).

Reply:
I actually don't have super strong feelings against adding metadata for no-op substitutions and no-match cases, except that a no-match could end up matching the wrong subset... Somewhat contrived example:

* subsets: {version=v1} {version=v2} {version=none}
* on_header_present rule to extract the version from `:path`, e.g.: `/.*\?version=(.*)`
* request comes in with `/none`

you end up routed to the `{version=none}` subset, whereas maybe your expectation was to hit the fallback policy. 

We could also improve `replaceAll()` to differentiate a no-op substitution versus no-match and treat those differently.... (the former adds metadata whereas the former doesn't?).

Reply:
I guess my example is stripping some prefix or suffix. So match `(.*)(_foo)?` and substitute `\1`. Then as you upgrade the downstream to not send the suffix, it still matches and converts to metadata but the value isn't actually changing.

Reply:
Ok -- let's allow no-substitution and no-match then.